pyomo/core/base/block.py

```diff
@@ -1139,6 +1140,14 @@ def del_component(self, name_or_object):
         #    return
 
         name = obj.local_name
+        if isinstance(obj, ComponentData) and not isinstance(obj, Component):
+            raise ValueError(
+                "Argument '%s' to del_component is a ComponentData object. Please "
+                "use the Python 'del' function to delete members of indexed "
+                "Pyomo objects. The del_component function can only be used to "
+                "delete IndexedComponents and ScalarComponents." % name
+            )
+
```

Comment:
This doesn't do what you think it does: line 1133 (`obj = self.component(name_or_object)`) maps the incoming object into its component, so you shouldn't be able to trigger this exception.  If we are going to change the behavior of `del_component` to only admit Components (either by name or reference), we should
- change (possibly inline the logic from) line 1133
- document this in the docstring
- possibly add a deprecation path (because we are changing behavior that has been in Pyomo for over a decade)??